docs/proposals/28-autonomous-shoot-clusters.md

```diff
@@ -0,0 +1,497 @@
+The configuration is specified via separate files.
+It needs a `Shoot` manifest and some additional configuration files, which usually reside in the garden cluster, e.g., a `CloudProfile` (best downloaded with `gardenadm discover`).
+For the initial cluster Gardener bootstrapping use-case however, these won't come from a pre-existing garden cluster as it doesn't exist yet.
+They can then be provided locally (that's why `gardenadm discover` and `gardenadm init` are separate commands), reflecting the to-be-deployed Gardener system later.
+
+The high-level steps executed by the command are as follows:
+
+1. Generate certificate authorities, certificates and keys using the [secrets manager](../development/secrets_management.md).
+2. Generate [`OperatingSystemConfig` resource](../extensions/operatingsystemconfig.md) (in memory) containing `systemd` units and files.
+   This includes the files for the minimal standard Kubernetes control plane, i.e., `kubelet`, `etcd`, `kube-apiserver`, `kube-controller-manager`, and `kube-scheduler`.
+   Also, it should be OS-agnostic (since there is no operating system config extension running anywhere yet).
+3. Reconcile the `OperatingSystemConfig` resource using the controller logic of [`gardener-node-agent`](../concepts/node-agent.md).
+4. `kubelet` starts and runs the static pods for the minimal standard Kubernetes control plane (`etcd`, `kube-apiserver`, `kube-controller-manager` and `kube-scheduler`).
+5. Deploy [`gardener-resource-manager`](../concepts/resource-manager.md) in [host network mode](#host-network-vs-pod-network).
+6. Apply the corresponding provider and networking extensions in [host network mode](#host-network-vs-pod-network).
+7. Deploy `kube-proxy` and CoreDNS for service routing and domain name resolution.
+8. Create a [`Network` resource](../extensions/network.md) and apply it to generate `ManagedResource` associated with the pod network.
+9. Apply the corresponding provider and networking extensions in [pod network mode](#host-network-vs-pod-network).
+   Also deploy the operating system config extensions.
+10. Redeploy `gardener-resource-manager` in [pod network mode](#host-network-vs-pod-network).
+11. Create the [`OperatingSystemConfig` resource](../extensions/operatingsystemconfig.md) now in the cluster and activate the `gardener-node-agent` process via its `systemd` unit.
+12. Apply [`ControlPlane` resource](../extensions/controlplane.md).
+
+##### Problems & Solutions
+
+As we intend to re-use code as much as possible, we need to jump through some hoops to make this work.
+
+###### Faking a Kubernetes API
+
+Most of the Gardener code base assumes an always existing garden cluster or the API server of a seed cluster, which might not be available during bootstrapping.
+A somewhat simple solution for components running as part of the same process is to use a fake client, e.g., `sigs.k8s.io/controller-runtime/pkg/client/fake` or `github.com/gardener/gardener/pkg/client/kubernetes/fake`.
```

Comment:
Yes, I agree. As discussed earlier today in the call, I guess it's fine to start with the fake client since we just rely on the simplest CREATE/UPDATE/GET functionality. In case we hit limits, we can easily implement our own client or follow the k3s approach.